### PR TITLE
Mobileapps 1997

### PR DESCRIPTION
**Pull request by swuforesee**
Implemented Android webview for SSO flow. Converted LoginController to a singleton.

To see a demo of the SSO flow, run FsrApiClient.AndroidSample. See testing instructions in the ticket for details.

### CODE REVIEW

File: acs-apiclient/Droid/WebView/ExternalFlowWebViewClient.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿using System.Threading;
+using Android.Webkit;
+using AcsApi;
+using Android.OS;
+
+namespace acs_apiclient.Droid
+{
+    
+}
```
Comment (ahusseini) [inline acs-apiclient/Droid/WebView/ExternalFlowWebViewClient.cs]: ❓ Should this be deleted?

File: acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                this.activity = activity;
+            }
+            
+            /**
+            * Use shouldOverrideUrlLoading(WebView, WebResourceRequest) instead when the Android app's min SDK is increased to >= 24
+            * https://developer.android.com/reference/android/webkit/WebViewClient.html#shouldOverrideUrlLoading(android.webkit.WebView, android.webkit.WebResourceRequest)
+            */
+            public override bool ShouldOverrideUrlLoading(WebView webview, String url)
+            {
+                Console.WriteLine($"ShouldOverrideUrlLoading: url={url}");
```
Comment (ahusseini) [inline acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs]: ⚠️ this testing console log should be removed

**Review by ahusseini — CHANGES_REQUESTED** [review]
⚠️ Looks good, with a couple of minor comments 

:bowtie: On a side note, it would be nice if we kept our lines within 120 characters.

File: acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    [Activity(Label = "ExternalFlowWebActivity", 
+    Theme = "@android:style/Theme.NoTitleBar", 
+    WindowSoftInputMode = SoftInput.AdjustResize)]
+    public class ExternalFlowWebActivity : Activity
+    {
+        public static string UrlParamKey = "url";
+        private static Android.Graphics.Color Obsidian = Android.Graphics.Color.ParseColor("#222b3c");
+        private static Android.Graphics.Color Ash = Android.Graphics.Color.ParseColor("#B7C0CC");
+        private static Android.Graphics.Color SelectedButtonTintColor = Ash;
+        private static Android.Graphics.Color UnselectedButtonTintColor = Obsidian;
```
Comment (sleeke) [inline acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs]: 💡 Can these be pulled into shared code?

File: acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            SetupSyncButton();
+            SetupWebView();
+        }
+
+        private void InitUri()
+        {
+            this.urlString = this.Intent.GetStringExtra(UrlParamKey);
+            if (String.IsNullOrEmpty(this.urlString))
+            {
+                throw new ArgumentException($"'url' must be provided in the Intent to start {typeof(ExternalFlowWebActivity)}");
```
Comment (sleeke) [inline acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs]: 💡 This should reference ```UrlParamKey``` directly so it is updated if it changes

Comment (sleeke) [reply acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs]: Also the error could be more specific:
```$"A URL must be provided for the '{UrlParamKey} key in the Intent to start {typeof(ExternalFlowWebActivity)}"```

File: acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs (hunk 2 above)
Comment (sleeke) [reply acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs]: Agreed

File: acs-apiclient/Droid/ExternalFlowWebActivity.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+using Android.Graphics;
+using System.Drawing;
+
+namespace acs_apiclient.Droid
+{
+    [Activity(Label = "ExternalFlowWebActivity", Theme = "@android:style/Theme.NoTitleBar")]
+    public class ExternalFlowWebActivity : Activity
+    {
+        public static string UrlParamKey = "url";
+        private static Android.Graphics.Color ObsidianGrey = Android.Graphics.Color.ParseColor("#222b3c");
```
Comment (sleeke) [inline acs-apiclient/Droid/ExternalFlowWebActivity.cs]: 💡 Probably a good candidate for shared code

File: acs-apiclient/Droid/ExternalFlowWebActivity.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            title.Text = newTitle;
+        }
+
+        private void SetupSyncButton()
+        {
+            EasyTintImageButton syncButton = FindViewById<EasyTintImageButton>(Resource.Id.easyTintImageButton_titlebarWithExitAndRefereshButtons_sync);
+            syncButton.SelectedTintColor = ObsidianGrey;
+            syncButton.Click += (object sender, EventArgs e) => 
+            {
+                //TODO implement behaviour after webview is setup
```
Comment (sleeke) [inline acs-apiclient/Droid/ExternalFlowWebActivity.cs]: ❓ Should this still be a todo? When will this be done?

File: acs-apiclient/Droid/ExternalFlowWebActivity.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+                Finish();
+            };
+        }
+
+        private void SetupWebView()
+        {
+            contentWebView = FindViewById<WebView>(Resource.Id.webview_externalFlowWeb_content);
+            contentWebView.Settings.JavaScriptEnabled = true;
+            contentWebView.SetWebViewClient(new ExternalFlowWebViewClient());
+            contentWebView.LoadUrl("http://www.google.com");
```
Comment (sleeke) [inline acs-apiclient/Droid/ExternalFlowWebActivity.cs]: ❓ Why is this needed?

File: acs-apiclient/Resources/layout/layout_titleBarWithExitAndSyncButtons.xml [hunk 8]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (sleeke) [inline acs-apiclient/Resources/layout/layout_titleBarWithExitAndSyncButtons.xml]: ⚠️ This would be more efficient with a frameLayout, or better yet (I think) a linearLayout - please ask me if you need a hand figuring out how to get the same result

File: acs-apiclient/Resources/values/Colors.xml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+﻿<?xml version="1.0" encoding="UTF-8" ?>
+<resources>
+    <color name="cloudy_blue">#B7C0CC</color>
+    <color name="ash">#B7C0CC</color>
+    <color name="obsidian">#222b3c</color>
```
Comment (sleeke) [inline acs-apiclient/Resources/values/Colors.xml]: 💡 It might be better in this case to set the colors in code - that way they can be shared more easily between iOS and Android

**Review by sleeke — CHANGES_REQUESTED** [review]
💡 Just a couple of suggestions here (see inline comments)

File: acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs (hunk 2 above)
Comment (swuforesee) [reply acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs]: Done.

File: acs-apiclient/Droid/WebView/ExternalFlowWebViewClient.cs (hunk 1 above)
Comment (swuforesee) [reply acs-apiclient/Droid/WebView/ExternalFlowWebViewClient.cs]: yup, deleted.

File: acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs (hunk 2 above)
Comment (swuforesee) [reply acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs]: Done.

File: acs-apiclient/Droid/WebView/ExternalFlowWebViewClient.cs (hunk 1 above)
Comment (swuforesee) [reply acs-apiclient/Droid/WebView/ExternalFlowWebViewClient.cs]: Removed.

File: acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs (hunk 3 above)
Comment (swuforesee) [reply acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs]: Pulled into new `Constants` class under `Colors`

File: acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs (hunk 4 above)
Comment (swuforesee) [reply acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs]: Done

File: acs-apiclient/Droid/ExternalFlowWebActivity.cs (hunk 5 above)
Comment (swuforesee) [reply acs-apiclient/Droid/ExternalFlowWebActivity.cs]: Pulled into `Constants.Colors`

File: acs-apiclient/Droid/ExternalFlowWebActivity.cs (hunk 6 above)
Comment (swuforesee) [reply acs-apiclient/Droid/ExternalFlowWebActivity.cs]: this code looks out of date....

It should be this:
`
private void SetupSyncButton()
        {
            EasyTintImageButton syncButton = FindViewById<EasyTintImageButton>(Resource.Id.easyTintImageButton_titlebarWithExitAndRefreshButtons_sync);
            syncButton.SelectedTintColor = SelectedButtonTintColor;
            syncButton.UnSelectedTintColor = UnselectedButtonTintColor;
            syncButton.Click += (object sender, EventArgs e) => 
            {
                this.contentWebView.LoadUrl(this.urlString);
            };
}
`
I'll look into this

File: acs-apiclient/Droid/ExternalFlowWebActivity.cs (hunk 7 above)
Comment (swuforesee) [reply acs-apiclient/Droid/ExternalFlowWebActivity.cs]: this is old code as well....

should be 
`
        private void SetupWebView()
        {
            this.contentWebView = FindViewById<WebView>(Resource.Id.webview_externalFlowWeb_content);
            this.contentWebView.Settings.JavaScriptEnabled = true;
            this.contentWebView.Settings.DomStorageEnabled = true;
            this.contentWebView.SetWebViewClient(new ExternalFlowWebViewClient(this));
            this.contentWebView.LoadUrl(this.urlString);
        }
`

Something weird is going on. This class looks completely out of date

File: acs-apiclient/Droid/ExternalFlowWebActivity.cs (hunk 6 above)
Comment (swuforesee) [reply acs-apiclient/Droid/ExternalFlowWebActivity.cs]: This class has been deleted. See the latest version of this class in acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs.

File: acs-apiclient/Droid/ExternalFlowWebActivity.cs (hunk 7 above)
Comment (swuforesee) [reply acs-apiclient/Droid/ExternalFlowWebActivity.cs]: This class has been deleted. See the latest version of this class in acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs.

File: acs-apiclient/Resources/layout/layout_titleBarWithExitAndSyncButtons.xml (hunk 8 above)
Comment (swuforesee) [reply acs-apiclient/Resources/layout/layout_titleBarWithExitAndSyncButtons.xml]: Done.

File: acs-apiclient/Resources/values/Colors.xml (hunk 9 above)
Comment (swuforesee) [reply acs-apiclient/Resources/values/Colors.xml]: Done.

**Comment by swuforesee** [comment]
Suggested improvements have been made. Please pull the latest version to get the latest login credentials that are automatically populated into the UI.

File: acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs (hunk 3 above)
Comment (sleeke) [reply acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs]: ✅

File: acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs (hunk 4 above)
Comment (sleeke) [reply acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs]: ✅

File: acs-apiclient/Droid/ExternalFlowWebActivity.cs (hunk 6 above)
Comment (sleeke) [reply acs-apiclient/Droid/ExternalFlowWebActivity.cs]: ✅

File: acs-apiclient/Droid/ExternalFlowWebActivity.cs (hunk 7 above)
Comment (sleeke) [reply acs-apiclient/Droid/ExternalFlowWebActivity.cs]: ✅

File: acs-apiclient/Resources/layout/layout_titleBarWithExitAndSyncButtons.xml (hunk 8 above)
Comment (sleeke) [reply acs-apiclient/Resources/layout/layout_titleBarWithExitAndSyncButtons.xml]: ✅

File: acs-apiclient/Resources/values/Colors.xml (hunk 9 above)
Comment (sleeke) [reply acs-apiclient/Resources/values/Colors.xml]: ✅

File: acs-apiclient/Droid/ExternalFlowWebActivity.cs (hunk 5 above)
Comment (sleeke) [reply acs-apiclient/Droid/ExternalFlowWebActivity.cs]: ✅

File: acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs (hunk 2 above)
Comment (sleeke) [reply acs-apiclient/Droid/Activity/ExternalFlowWebActivity.cs]: ✅ Not done, but fixed in review commit

**Comment by swuforesee** [comment]
@ahmadhusseini it looks like both you and @sleeke have reviewed the code and i've also made changes. 

Can you take a look through to see if there is anything else that is missing? I know this is blocking you from moving forward because of my change to the LoginController into a singleton.
